server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -287,6 +287,19 @@ static void parseObjectOrNested(DocumentParserContext context) throws IOExceptio
         }
 
         if (context.parent().isNested()) {
+            if (context.mappingLookup().isSourceSynthetic() && context.getClonedSource() == false) {
```

Comment:
In that case is the `|| mapper instanceof NestedObjectMapper` addition at line 612 required?

Reply:
It is, that section covers the usual case where the nested object  contains an array.

Reply:
If a nested object has already been added to ignored source, then I think we don't need to do anything for that json object contains an array? Since it has already been added to ignored source as part of the object?

Reply:
I believe we either get an array or an object, not an object wrapping an array. I verified this in the unit test, isn't that the valid syntax.

Reply:
I see without `|| mapper instanceof NestedObjectMapper` we only store the first element of an array. So subsequent objects in an array would not get recorded.

Reply:
Done.